### PR TITLE
Optimize balance updates

### PR DESCRIPTION
**Pull request by antonyshchenko**
Prior to this PR to store balance updates we were:
- fetching starting balances from DB
- making the insert or update in the DB
- fetching updated balances from DB to generated Broadcaster notifications

In this PR:
- Balance updates are done with a single DB call
- Broadcaster notifications for balance updates are generated without an aditional database query
- Websocket client receives all balances on subscription and then only the updated balances

### CODE REVIEW

File: integrationtests/src/test/kotlin/xyz/funkybit/integrationtests/api/OrderExecutionTest.kt [hunk 1]
```diff
@@ -716,13 +710,13 @@ class OrderExecutionTest : OrderBaseTest() {
             assertDbCallsCountInTrace(12, applyOrderBatchTraceIds[1])
             when (chainIndex) {
                 0 -> {
-                    assertDbCallsCountInTrace(26, applyOrderBatchTraceIds[2])
-                    assertDbCallsCountInTrace(25, applyOrderBatchTraceIds[3])
+                    assertDbCallsCountInTrace(23, applyOrderBatchTraceIds[2])
```
Comment (antonyshchenko) [inline integrationtests/src/test/kotlin/xyz/funkybit/integrationtests/api/OrderExecutionTest.kt:713]: Reduced the number of DB calls

File: backend/src/main/kotlin/xyz/funkybit/core/model/db/migrations/V117_AddUniqueConstraintToBalanceTable.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    object V117_BalanceTable : GUIDTable<BalanceId>("balance", ::BalanceId) {
+        val symbolGuid = reference("symbol_guid", V117_SymbolTable).index()
+        val walletGuid = reference("wallet_guid", V117_WalletTable).index()
+        val type = customEnumeration(
+            "type",
+            "BalanceType",
+            { value -> V117_BalanceType.valueOf(value as String) },
+            { PGEnum("BalanceType", it) },
+        )
```
Comment (imykolenko) [inline backend/src/main/kotlin/xyz/funkybit/core/model/db/migrations/V117_AddUniqueConstraintToBalanceTable.kt:30]: those fields seems to be already existing, is it necessary to list them?

File: backend/src/main/kotlin/xyz/funkybit/core/model/db/Balance.kt [hunk 3]
```diff
@@ -120,85 +121,92 @@ class BalanceEntity(guid: EntityID<BalanceId>) : GUIDEntity<BalanceId>(guid) {
...
+                        ${BalanceTable.createdAt.name},
+                        ${BalanceTable.updatedAt.name},
+                        ${BalanceTable.createdBy.name},
+                        ${BalanceTable.updatedBy.name}
+                    ) VALUES 
+                        ${(assignments.indices).joinToString(",\n") { "(typeid_generate_text('balance'), ?, ?, ?, ?, ?, ?, ?, ?)" }}
+                    ON CONFLICT (${BalanceTable.symbolGuid.name}, ${BalanceTable.walletGuid.name}, ${BalanceTable.type.name})
+                    DO UPDATE SET 
+                        ${BalanceTable.balance.name} = ${
+                    if (replace) {
```
Comment (imykolenko) [inline backend/src/main/kotlin/xyz/funkybit/core/model/db/Balance.kt:161]: 👍

**Review by imykolenko — APPROVED** [review]
Looks great!

File: backend/src/main/kotlin/xyz/funkybit/core/model/db/migrations/V117_AddUniqueConstraintToBalanceTable.kt (hunk 2 above)
Comment (antonyshchenko) [reply backend/src/main/kotlin/xyz/funkybit/core/model/db/migrations/V117_AddUniqueConstraintToBalanceTable.kt:30]: I had to define them so I could reference them in the unique constraint definition

File: backend/src/main/kotlin/xyz/funkybit/core/model/db/Balance.kt [hunk 4]
```diff
@@ -120,85 +121,92 @@ class BalanceEntity(guid: EntityID<BalanceId>) : GUIDEntity<BalanceId>(guid) {
...
+                        assignments.flatMap {
+                            listOf(
+                                BalanceTable.symbolGuid.columnType.valueToDB(it.symbolId),
+                                BalanceTable.walletGuid.columnType.valueToDB(it.walletId),
+                                BalanceTable.type.columnType.valueToDB(it.balanceType),
+                                BalanceTable.balance.columnType.valueToDB(it.value.toBigDecimal()),
+                                BalanceTable.createdAt.columnType.valueToDB(timestamp),
+                                BalanceTable.updatedAt.columnType.valueToDB(timestamp),
+                                BalanceTable.createdBy.columnType.valueToDB("system"),
+                                BalanceTable.updatedBy.columnType.valueToDB("system"),
```
[the remark this answers is not in the document]
Comment (antonyshchenko) [reply backend/src/main/kotlin/xyz/funkybit/core/model/db/Balance.kt]: changed

**Pull request merged**